stan/math/opencl/prim/lub_constrain.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_MATH_OPENCL_PRIM_LUB_CONSTRAIN_HPP
+#define STAN_MATH_OPENCL_PRIM_LUB_CONSTRAIN_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the lower and upper-bounded matrix derived by
+ * transforming the specified free matrix given the specified
+ * lower and upper bounds.
+ *
+ * <p>The transform is the transformed and scaled inverse logit,
+ *
+ * <p>\f$f(x) = L + (U - L) \mbox{logit}^{-1}(x)\f$
+ *
+ * @tparam T matrix expression type
+ * @tparam L lower bound expression type
+ * @tparam U upper bound expression type
+ * @param[in] x Free matrix to transform.
+ * @param[in] lb Lower bound
+ * @param[in] ub Upper bound
+ * @return Lower- and upper-bounded matrix derived from transforming
+ *   the free matrix.
+ * @throw std::domain_error if ub <= lb
+ */
+template <typename T, typename L, typename U,
+          require_all_kernel_expressions_and_none_scalar_t<T>* = nullptr,
+          require_all_kernel_expressions_t<L, U>* = nullptr>
+inline matrix_cl<double> lub_constrain(const T& x, const L& lb, const U& ub) {
+  auto diff = ub - lb;
+  auto lb_inf = lb == NEGATIVE_INFTY;
+  auto ub_inf = ub == INFTY;
+
+  auto check
+      = check_cl("lub_constrain (OpenCL)", "(ub - lb)", diff, "positive");
+  matrix_cl<double> res;
+
+  results(check, res) = expressions(
```

Comment:
Kernel uses an extra buffer for check results. There it stores if any check should fail. After the kernel is launched, it is waited on and these results are checked. If necessary an exception is thrown. 

Interface looks a bit weird, but that is the best I came up with so that checks can be done in the same kernel as calculations.